http/src/main/scala/org/http4s/blaze/http/ClientSessionManagerImpl.scala

```diff
@@ -0,0 +1,167 @@
+     UrlComposition(request.url) match {
+      case Success(urlComposition) =>
+        val id = getId(urlComposition)
+        val session = findExistingSession(id)
+
+        if (session == null) createNewSession(urlComposition, id)
+        else {
+          logger.debug(s"Found hot session for id $id: $session")
+          Future.successful(session)
+        }
+
+      case Failure(t) => Future.failed(t)
+    }
+  }
+
+  // WARNING: can emit `null`. For internal use only
+  private[this] def findExistingSession(id: ConnectionId): HttpClientSession = sessionCache.synchronized {
+    sessionCache.get(id) match {
+      case null => null // nop
+      case sessions =>
+        var session: HttpClientSession = null
+        val it = sessions.iterator
+        while(session == null && it.hasNext) it.next() match {
+          case h2: Http2ClientSession if h2.status == Closed =>
+            h2.closeNow() // make sure its closed
+            it.remove()
+
+          case h2: Http2ClientSession if h2.quality > 0.1 =>
+            session = h2
+
+          case _: Http2ClientSession => () // nop
+
+          case h1: Http1ClientSession =>
+            it.remove()
+            if (h1.status != Closed) { // Should never be busy
+              session = h1 // found a suitable HTTP/1.x session.
+            }
+        }
+
+        // if we took the last session, drop the collection
+        if (sessions.isEmpty) {
+          sessionCache.remove(id)
+        }
+
+        session
+    }
+  }
+
+  private[this] def createNewSession(urlComposition: UrlComposition, id: ConnectionId): Future[HttpClientSession] = {
+    logger.debug(s"Creating new session for id $id")
+    val p = Promise[HttpClientSession]
+
+    socketFactory.connect(urlComposition.getAddress).onComplete {
+      case Failure(e) => p.tryFailure(e)
+      case Success(head) =>
+        val clientStage = new Http1ClientStage(config)
+        var builder = LeafBuilder(clientStage)
+        if (urlComposition.scheme == "https") {
+          val engine = config.getClientSslEngine()
+          engine.setUseClientMode(true)
+          builder = builder.prepend(new SSLStage(engine))
+        }
+
+        builder.base(head)
+        head.sendInboundCommand(Command.Connected)
+        p.trySuccess(new Http1SessionProxy(id, clientStage))
+      }
+
+    p.future
+  }
+
+  /** Return the session to the pool.
+    *
+    * Depending on the state of the session and the nature of the pool, this may
+    * either cache the session for future use or close it.
+    */
+  override def returnSession(session: HttpClientSession): Unit = {
+    logger.debug(s"Returning session $session")
+    session match {
+      case _ if session.isClosed => () // nop
+      case _: Http2ClientSession => () // nop
+      case h1: Http1ClientSession if !h1.isReady =>
+        logger.debug(s"Closing unready session $h1")
+        h1.closeNow() // we just orphan the Future. Don't care.
```

Comment:
Are you throwing away an exception if it fails?

Reply:
Yes, I didn't think it would be valuable. I can log on the failure pathway if you like: could be interesting to know.

Reply:
Turning `Future[Unit]` into `Unit` is essentially an empty catch block.  I have been burned rarely, but badly, by this.  If you're going to discard it, at least log the failures.

(I wouldn't have noticed it if not for your comment.)